text/0000-replace-ref-by-asterisk.md

```diff
@@ -0,0 +1,120 @@
+- Start Date: 2015-01-27
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+Change `ref` in patterns to `*`, to make pattern syntax better match expression syntax.
+Example: `Some(ref x)` becomes `Some(*x)`.
+This change removes all the usage of `ref` keyword,
+so the RFC also discusses the fate of it.
+
+# Motivation
+
+In current design, pattern syntax tries to mirror expression syntax
+(with some unavoidable `mut` related exceptions), so one may think that
+"everything works backwards". But `ref` is an exception here, because:
+
+* It doesn't appear in expression syntax at all.
+* It works "forwards" in contrary to rest of pattern syntax.
+
+Such an exception may be confusing for a newcommer and even
+some more experienced Rust programmer may feel like `ref` doesn't fit pattern syntax well.
+Additionaly, one can read both `ref` and `&` as *ref*, but they serve totally different purposes.
+That is another source of confusion.
+
+This RFC proposes to change `ref` to `*` to make this exception disappear.
+
+Following example shows the symmetry of patterns and expressions in proposed design.
+Notice that
+the syntax used to reconstruct `x` value is exactly the same as the syntax used to match on in.
+
+```rust
+let x = Some(5);
+if let Some(*y) = x {
+    assert_eq!(Some(*y), x)
+}
+```
+
+Another motivation are opinions, heard from time to time, that `ref` should be called `deref`.
+
+Yet another reason for this change is discoverability -- it's reusing a operator
+(while mostly keeping its semantics),
+so it's not unlikely for somebody to discover this syntax without consulting a manual.
+Actually, there are some users reporting that they tried to use `*` before learning about `ref`.
+
+# Detailed design
+
+### Interaction with `mut`
+
+This RFC proposes to change:
+
+```rust
+ref x            into       *x
+ref mut x        into       mut *x
+mut ref x        into       * mut x
+mut ref mut x    into       mut * mut x
+```
+
+The rest of pattern syntax stays unchanged.
+Notice that in addition to simply changing one token to another,
+this RFC also proposes to change placing of `mut` keywords.
+The reason for that change is to make `mut` refer to the
+thing on the right of `mut`:
+
+* `mut` keyword in `mut *x` means "dereference of `x` is mutable", because it parses as `mut (*x)`,
+* `mut` keyword in `* mut x` means "`x` is mutable reference", because it parses as `*(mut x)`.
+
+Syntax may allow these parenthesised versions too, for consistency / increase of readability.
+
+Note that most common usage of `mut` would be `mut *x` (corresponding to current `ref mut x`),
+so the raw-pointer-resembling `* mut` wouldn't appear very often
+(in rust source `ref mut` appears 139 times, compared to only 9 usages of `mut ref`).
```

Comment:
Furthermore, none of them are equivalent to `let mut x: &T = &val;`: `mut ref <ident>` isn't a valid pattern. (They're all parsed as `&mut (ref ...)`, i.e. pattern matching through a `&mut`.)